### PR TITLE
Dstar updates for behavior

### PR DESCRIPTION
**Pull request by Isaac-F24**
Changes Dstar node into a class, refactors some of the old / unneeded parts of Dstar/dstar node, and (crucial) allows dstar to follow goals that are out of range of the currently generated occupancy map.

Want to make sure that this doesn't break anything for anybody.

### CODE REVIEW

**Comment by ashviniyer21** [comment]
Can you clarify what the behavior looks like if planning to something outside of the occupancy range? Does it treat everything outside of the occupancy range as a "free space?"

File: lunabot_nav/scripts/dstar_node.py [hunk 1]
```diff
@@ -9,157 +9,224 @@
...
-    global update_grid, grid, res, x_offset, y_offset
-    width = data.info.width
-    height = data.info.height
-    grid = np.reshape(data.data, (height, width))
-    res = data.info.resolution
-    x_offset = data.info.origin.position.x
-    y_offset = data.info.origin.position.y
-    update_grid = True
+        self.grid_update_needed: bool = False
+        self.goal_update_needed: bool = False
```
Comment (ashviniyer21) [inline lunabot_nav/scripts/dstar_node.py:20]: Repeated line?

File: lunabot_nav/scripts/dstar_node.py [hunk 2]
```diff
@@ -9,157 +9,224 @@
...
-    index = 0
-    for i in range(data.y, data.y + data.height):
-        for j in range(data.x, data.x + data.width):
-            temp_map[i][j] = data.data[index]
-            index += 1
+        self.dstar: Dstar = None
 
-    grid = temp_map.copy()
-    update_grid = True
+        self.path_sampling_rate = 5 # Take every <n-th> point from the path
```
Comment (ashviniyer21) [inline lunabot_nav/scripts/dstar_node.py:29]: Lets make this a rosparam later (I personally don't like constants hidden inside files like this, I would at the very least move it to the top of the function)

File: lunabot_nav/scripts/dstar_node.py [hunk 3]
```diff
@@ -9,157 +9,224 @@
...
-        # print("Loop")
-        if (dstar is None) and len(grid) > 0 and len(pose) > 0 and len(goal) > 0:
-            dstar = Dstar(goal, pose, grid, radius, res, x_offset, y_offset)
-            print("Initialize")
+            columns = np.ones((len(self.grid), difference)) * -1
+            self.grid = np.concatenate((columns, self.grid), axis=1)
 
-        if dstar is not None:
-            # print("Iterate")
+            # change offset
```
Comment (ashviniyer21) [inline lunabot_nav/scripts/dstar_node.py:111]: How is this logic handled when a new map update occurs? I don't see any logic about x_offset being fixed when updating map so will it override the wrong cells in the map?

Comment (ashviniyer21) [reply lunabot_nav/scripts/dstar_node.py:111]: Same issue for `y_axis` code update above

File: lunabot_nav/src/lunabot_nav/dstar.py [hunk 4]
```diff
@@ -97,6 +89,7 @@ def insert(self, node: list, key: tuple):
         self.node_queue.put((key, node))
 
     # removes a given node from the queue
+    # TODO- this probably is not efficient
```
Comment (ashviniyer21) [inline lunabot_nav/src/lunabot_nav/dstar.py:92]: You could replace the priority queue with a self-balanced BST to get log n removal time. I dont see an easy python library for this, and I don't think its a big deal (Dstar instead really our bottleneck here)

**Comment by Isaac-F24** [comment]
> Can you clarify what the behavior looks like if planning to something outside of the occupancy range? Does it treat everything outside of the occupancy range as a "free space?"

right now yes, everything not currently mapped is free space
Also, this definitely doesn't work yet

**Review by ashviniyer21 — COMMENTED** [review]
Left a few minor comments. Main thing that concerns me is your logic for expanding the grid; I think there might be some issues when the grid gets re-initialized potentially.

File: lunabot_nav/scripts/dstar_node.py (hunk 3 above)
Comment (Isaac-F24) [reply lunabot_nav/scripts/dstar_node.py:111]: Yes this is the problem

File: lunabot_nav/scripts/dstar_node.py (hunk 1 above)
Comment (Isaac-F24) [reply lunabot_nav/scripts/dstar_node.py:20]: not quite

Comment (ashviniyer21) [reply lunabot_nav/scripts/dstar_node.py:20]: Can you clarify?

Comment (ashviniyer21) [reply lunabot_nav/scripts/dstar_node.py:20]: Oh im blind, never mind lol

File: lunabot_config/config/global_costmap_sim.yml [hunk 5]
```diff
@@ -7,6 +7,6 @@ robot_base_frame: "base_link"
 footprint: "[ [0.54615446154,0.295590],[0.546154,-0.295590],[-0.546154,-0.295590],[-0.546154,0.295590] ]"
 map_topic: "/projected_map"
 inflation: {
-  inflation_radius: 0.4,
-  cost_scaling_factor: 10,
+  inflation_radius: 0.6,
```
Comment (ashviniyer21) [inline lunabot_config/config/global_costmap_sim.yml:10]: Why are these parameters changed? Just for testing or are these actually better?

File: lunabot_nav/scripts/dstar_node.py [hunk 6]
```diff
@@ -9,157 +9,175 @@
...
-        if (dstar is None) and len(grid) > 0 and len(pose) > 0 and len(goal) > 0:
-            dstar = Dstar(goal, pose, grid, radius, res, x_offset, y_offset)
-            print("Initialize")
+        odom_topic = rospy.get_param("/odom_topic")
+        goal_topic = rospy.get_param("/nav_goal_topic")
 
-        if dstar is not None:
-            # print("Iterate")
+        radius = 8  # robot rad (grid units)
+        # TODO why was this here
```
Comment (ashviniyer21) [inline lunabot_nav/scripts/dstar_node.py:90]: TODO still here lol. Should probably address this

File: lunabot_nav/src/lunabot_nav/dstar.py [hunk 7]
```diff
@@ -1,92 +1,105 @@
...
-        self.node_values_list = maxsize * np.ones(
-            (init_map.shape[0], init_map.shape[1], 2)
-        )  # 2d Array of nodes: each value is [Distance (g), Estimate (rhs)]
+        self.node_queue: PriorityQueue = PriorityQueue()
+        self.km: float = 0.0
 
-        self.x_offset = x_offset
-        self.y_offset = y_offset
+        # What number on the map corresponds to occupied
+        self.OCCUPANCY_THRESHOLD = 50
```
Comment (ashviniyer21) [inline lunabot_nav/src/lunabot_nav/dstar.py:31]: This is already a rosparam?

File: lunabot_nav/src/lunabot_nav/dstar.py [hunk 8]
```diff
@@ -304,337 +239,347 @@ def update_node(self, node: list, init: bool):
...
+        # After all done updating, calculate the new path
+        return self.find_path()
+
+    def buffer_map_for_goal(self):
+        """
+        Add a buffer to the map so we can plan to the goal when it is outside the map.
+        Saves the amount of buffer to allow for correct translations between real-world coords and grid coords.
+        """
+
+        EXTRA_BUFFER = 3  # extra buffer to add to the map, needed because the real-world coordinates can be slightly off
```
Comment (ashviniyer21) [inline lunabot_nav/src/lunabot_nav/dstar.py:447]: How was the value `3` picked? Will this need to be changed in the future? Imo it makes more sense for this to be a factor of resolution

File: lunabot_nav/src/lunabot_nav/dstar.py [hunk 9]
```diff
@@ -304,337 +239,347 @@ def update_node(self, node: list, init: bool):
...
+        self.current_node[1] += columns_left_node_values
+
+        self.node_values_list = new_node_values
+
+        self.current_map = new_map
 
-        # New path will be published
-        # path = self.createPathList()
+        # Call update-replan, which compares the prev map and new map to mark any differences.
+        # This eventually calcualtes the new path and returns it.
```
Comment (ashviniyer21) [inline lunabot_nav/src/lunabot_nav/dstar.py:584]: nit: typo lol

File: lunabot_nav/src/lunabot_nav/dstar.py [hunk 10]
```diff
@@ -1,92 +1,105 @@
...
+        self.prev_node: list[int] = start
 
-    # test
+        # Insert the goal node into the priority queue
+        self.insert(self.goal, self.calculate_key(self.goal))
 
-    # Update Pose (current position)
     def update_position(self, coords):
-        position = self.convertToGrid(coords)
+        """
```
Comment (ashviniyer21) [inline lunabot_nav/src/lunabot_nav/dstar.py:72]: You only expand the buffer at the beginning for the goal. What happens if the robot traverses into a space where the buffer needs to be expanded further? I don't think thats addressed in these changes

**Review by ashviniyer21 — CHANGES_REQUESTED** [review]
Changes look good for the most part, main thing that concerns me is if the robot ever travels to a location outside of the current map thats not accounted for (only expanding the map for the goal is) Added some other minor comments, main thing I want addressed is this though

File: lunabot_nav/src/lunabot_nav/dstar.py (hunk 10 above)
Comment (ashviniyer21) [reply lunabot_nav/src/lunabot_nav/dstar.py:72]: Nevermind, didn't see the buffer gets updated in `update_map`. I'm still a bit concerned about timing w/ `update_pos` (if `update_pos` gets called before `update_map`, then may run into an issue where you try setting the pos as a negative grid index)

Comment (ashviniyer21) [reply lunabot_nav/src/lunabot_nav/dstar.py:72]: Looked at this further / discussed, it should be ok

File: lunabot_config/config/global_costmap_sim.yml (hunk 5 above)
Comment (Isaac-F24) [reply lunabot_config/config/global_costmap_sim.yml:10]: These are better as right now, without it, the robot gets too close to obstacles

File: lunabot_nav/src/lunabot_nav/dstar.py (hunk 8 above)
Comment (Isaac-F24) [reply lunabot_nav/src/lunabot_nav/dstar.py:447]: randomly- i suspect 1 might have worked also but i gave it the extra just in case.

**Pull request merged**